core/services/p2p/peer.go

```diff
@@ -0,0 +1,250 @@
+package p2p
+
+import (
+	"context"
+	"crypto/ed25519"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/smartcontractkit/libocr/networking/ragedisco"
+	"github.com/smartcontractkit/libocr/ragep2p"
+	ragetypes "github.com/smartcontractkit/libocr/ragep2p/types"
+
+	nettypes "github.com/smartcontractkit/libocr/networking/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+
+	commonlogger "github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+var (
+	defaultGroupID    = [32]byte{0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01}
+	defaultStreamName = "stream"
+)
+
+type Peer interface {
+	services.Service
+	UpdatePeerAddresses(peerIDs []ragetypes.PeerID) error
+	Send(peerID ragetypes.PeerID, msg []byte) error
+	Receive() <-chan Message
```

Comment:
Is there a benefit to exposing the internal detail of how reception happens (i.e. via a channel)? As opposed to exposing a generic Receive(Message) function?

Reply:
I would bet the channel makes things more difficult if you are mapping to grpc streams underneath.

Reply:
`chan` seems fine for a raw API though :shrug:

Reply:
By "generic Receive(Message) function?" do you mean a callback? I want to push all messages to some Receiver/Dispatcher object as they arrive, channel seems fine to me.

Reply:
Yes, I mean hiding the channel as an internal detail as opposed to having it as an external one. One benefit for exposing the channel I guess is that you can use it for synchronization with `select`.